indispensable/src/commonMain/kotlin/at/asitplus/signum/indispensable/symmetric/SymmetricKey.kt

```diff
@@ -0,0 +1,211 @@
+     */
+    sealed class Integrated<A : AuthCapability<KeyType.Integrated>, I : NonceTrait>
+    @HazardousMaterials("Does not check whether key size matched algorithm! Useful for testing, but not production!")
+
+    constructor(
+        override val algorithm: SymmetricEncryptionAlgorithm<A, I, KeyType.Integrated>,
+        /**
+         * The actual encryption key bytes
+         */
+        val secretKey: ByteArray
+    ) :
+        SymmetricKey<A, I, KeyType.Integrated> {
+        sealed class Authenticating<I : NonceTrait>(
+            algorithm: SymmetricEncryptionAlgorithm<AuthCapability.Authenticated.Integrated, I, KeyType.Integrated>,
+            secretKey: ByteArray
+        ) : Integrated<AuthCapability.Authenticated.Integrated, I>(algorithm, secretKey),
+            SymmetricKey.Authenticating<AuthCapability.Authenticated.Integrated, I, KeyType.Integrated> {
+
+            class RequiringNonce(
+                algorithm: SymmetricEncryptionAlgorithm<AuthCapability.Authenticated.Integrated, NonceTrait.Required, KeyType.Integrated>,
+                secretKey: ByteArray,
+                override val additionalProperties: MutableMap<String, String> = mutableMapOf<String, String>()
+            ) : Authenticating<NonceTrait.Required>(
+                algorithm, secretKey
+            ), SymmetricKey.RequiringNonce<AuthCapability.Authenticated.Integrated, KeyType.Integrated>
+
+            class WithoutNonce(
+                algorithm: SymmetricEncryptionAlgorithm<AuthCapability.Authenticated.Integrated, NonceTrait.Without, KeyType.Integrated>,
+                secretKey: ByteArray,
+                override val additionalProperties: MutableMap<String, String> = mutableMapOf<String, String>()
+            ) : Authenticating<NonceTrait.Without>(
+                algorithm, secretKey
+            ), SymmetricKey.WithoutNonce<AuthCapability.Authenticated.Integrated, KeyType.Integrated>
+        }
+
+        sealed class NonAuthenticating<I : NonceTrait>(
+            algorithm: SymmetricEncryptionAlgorithm<AuthCapability.Unauthenticated, I, KeyType.Integrated>,
+            secretKey: ByteArray
+        ) : Integrated<AuthCapability.Unauthenticated, I>(algorithm, secretKey), SymmetricKey.NonAuthenticating<I> {
+            class RequiringNonce(
+                algorithm: SymmetricEncryptionAlgorithm<AuthCapability.Unauthenticated, NonceTrait.Required, KeyType.Integrated>,
+                secretKey: ByteArray,
+                override val additionalProperties: MutableMap<String, String> = mutableMapOf<String, String>()
+            ) : NonAuthenticating<NonceTrait.Required>(algorithm, secretKey)
+
+            class WithoutNonce(
+                algorithm: SymmetricEncryptionAlgorithm<AuthCapability.Unauthenticated, NonceTrait.Without, KeyType.Integrated>,
+                secretKey: ByteArray,
+                override val additionalProperties: MutableMap<String, String> = mutableMapOf<String, String>()
```

Comment:
why is the same `additionalProperties` `override val` declared in every single ctor? shouldn't the val be further down in this case?

Reply:
those are the leaves

Reply:
yes, why is it not on a superclass if every subclass has the same member